tasks/linter_codegen/src/main.rs

```diff
@@ -0,0 +1,330 @@
+        if parts.len() != 2 {
+            continue;
+        }
+        let Some(plugin_module_name) = parts.first() else { continue };
+        let Some(rule_module_name) = parts.get(1) else { continue };
+        rule_entries.push(RuleEntry { plugin_module_name, rule_module_name });
+    }
+    // Sort deterministically
+    rule_entries.sort_by(|a, b| {
+        let ord = a.plugin_module_name.cmp(b.plugin_module_name);
+        if ord == std::cmp::Ordering::Equal {
+            a.rule_module_name.cmp(b.rule_module_name)
+        } else {
+            ord
+        }
+    });
+
+    Ok(rule_entries)
+}
+
+/// Detect the top-level node types used in a lint rule file by analyzing the Rust AST with `syn`.
+/// Returns `Some(bitset)` if at least one node type can be determined, otherwise `None`.
+fn detect_top_level_node_types(file: &File, rule: &RuleEntry) -> Option<BTreeSet<String>> {
+    let rule_impl = find_rule_impl_block(file, &rule.rule_struct_name())?;
+    let run_func = find_impl_function(rule_impl, "run")?;
+
+    let variants: BTreeSet<String> = if let Some(det) = IfElseKindDetector::from_run_func(run_func)
+    {
+        det.variants
+    } else {
+        return None;
+    };
+    if variants.is_empty() {
+        return None;
+    }
+
+    Some(variants)
+}
+
+fn find_rule_impl_block<'a>(file: &'a File, rule_struct_name: &str) -> Option<&'a syn::ItemImpl> {
+    for item in &file.items {
+        let syn::Item::Impl(imp) = item else { continue };
+        let ident = match imp.self_ty.as_ref() {
+            syn::Type::Path(p) => p.path.get_ident(),
+            _ => None,
+        };
+        if ident.is_some_and(|id| id == rule_struct_name)
+            && imp.trait_.as_ref().is_some_and(|(_, path, _)| path.is_ident("Rule"))
+        {
+            return Some(imp);
+        }
+    }
+    None
+}
+
+fn find_impl_function<'a>(imp: &'a syn::ItemImpl, func_name: &str) -> Option<&'a syn::ImplItemFn> {
+    for impl_item in &imp.items {
+        let syn::ImplItem::Fn(func) = impl_item else { continue };
+        if func.sig.ident == func_name {
+            return Some(func);
+        }
+    }
+    None
+}
+
+/// Detects top-level `if let AstKind::... = node.kind()` patterns in the `run` method.
+struct IfElseKindDetector {
+    variants: BTreeSet<String>,
+}
+
+impl IfElseKindDetector {
+    fn from_run_func(run_func: &syn::ImplItemFn) -> Option<Self> {
+        // Only consider when the body has exactly one top-level statement and it's an `if`.
+        let block = &run_func.block;
+        if block.stmts.len() != 1 {
+            return None;
+        }
+        let stmt = &block.stmts[0];
+        let Stmt::Expr(Expr::If(ifexpr), _) = stmt else { return None };
+        let mut variants = BTreeSet::new();
+        let result = collect_if_chain_variants(ifexpr, &mut variants);
+        if result == CollectionResult::Incomplete || variants.is_empty() {
+            return None;
+        }
```

Comment:
The pattern of passing `out: &mut BTreeSet<String>` through various free functions is not incorrect, but it's perhaps a little un-idiomatic. I wonder if it'd be preferable to make some of these functions methods of the `IfElseKindDetector` trait instead?

Or, if you want to share some of the logic with other detector structs in future PRs, add a `Variants` type which wraps a `BTreeSet<String>` and has methods like `extract_variants_from_pat` on it?

As I said in an earlier comment, the problem with `syn` is that it's really verbose and hard to follow (or at least I find it so). So in my view, anything we can do to make this code as structured and clear as possible will be a big help with verifying correctness.